std/machines/hash/poseidon2_bb.asm

```diff
@@ -0,0 +1,194 @@
+    // that makes up for the input field element. 
+    // For input i, we expect the field element at address input_addr + 4 * i.
+    // TODO: This could be an intermediate polynomial, but for some reason estark-starky
+    // fails then, so we keep it as a witness for now
+    let do_mload;
+    do_mload = used * sum(STATE_SIZE, |i| CLK[i]);
+    let input_index = sum(STATE_SIZE, |i| expr(i) * CLK[i]);
+    link if do_mload ~> (word_high, word_low) = mem.mload(input_addr + 4 * input_index, time_step);
+
+    // Combine the low and high limbs and write it into `input`
+    let current_input = array::sum(array::new(STATE_SIZE, |i| CLK[i] * input[i]));
+    do_mload * (word_low + word_high * 2**16 - current_input) = 0;
+
+    // Do a memory write in each of the next OUTPUT_SIZE rows.
+    // For output i, we write the two limbs of field element at address output_addr + 4 * i.
+    let do_mstore = used * sum(OUTPUT_SIZE, |i| CLK[i + STATE_SIZE]);
+    let output_index = sum(OUTPUT_SIZE, |i| expr(i) * CLK[i + STATE_SIZE]);
+    // TODO: This translates to two additional permutations. But because they go to the same machine
+    // as the mloads above *and* never happen at the same time, they could actually be combined with
+    // the mload permutations. But there is currently no way to express this.
+    link if do_mstore ~> mem.mstore(output_addr + 4 * output_index, time_step + 1, word_high, word_low);
+
+    // Make sure that in row i + STATE_SIZE, word_low and word_high correspond to output i
+    let current_output = array::sum(array::new(OUTPUT_SIZE, |i| CLK[i + STATE_SIZE] * output[i]));
+    link if do_mstore ~> (word_low, word_high) = split_bb.split(current_output);
+
+
+    // ------------- End memory read / write ---------------
+
+    pol constant LASTBLOCK(i) { if i % ROWS_PER_HASH == ROWS_PER_HASH - 1 { 1 } else { 0 } };
+    // Like LASTBLOCK, but also 1 in the last row of the table
+    // Specified this way because we can't access the degree in the match statement
+    pol constant LAST = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]* + [1];
+
+    // Whether the current round is a partial round
+    pol constant PARTIAL = [0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0]*;
+    
+    // The round constants
+    pol constant C_0 = [0xb585f766f2144405, 0x86287821f722c881, 0xe9fa634a21de0082, 0x92a756e67e2b9413, 0x3cc3f892184df408, 0x7131aa45268d7d8c, 0x99ad1aab0814283b, 0xeb84f608da56ef48, 0x7159cd30c3ac118e, 0xdcef0797c2b69ec7, 0xd0762cbc8ca6570c, 0x30a4680593258387, 0x15a16a8a8322d458, 0x5a3f1bb1c53a9645, 0x775005982d74d7f7, 0xf9cc95c22b4c1fcc, 0xc49366bb25e8513, 0xdd611f1000c17442, 0x2ff876fa5ef97c4, 0x3d06c8bd1514e2d9, 0xe89cd854d5d01d33, 0xece5a71e0cfedc75, 0x90004c1371b893c5, 0xde122bebe9a39368, 0x4d61e56a525d225a, 0x1478d361dbbf9fac, 0x475cd3205a3bdcde, 0xe70201e960cb78b8, 0x7be5b9ffda905e1c, 0xf3c12fe54d5c653b, 0x0]*;
```

Comment:
Can these be shared across the different poseidon implementations?

Reply:
No, because the number of rounds depends on the state length, which depends on the field size. Also these constants don't fit in BabyBear.